src/nanotron/fp8/tensor.py

```diff
@@ -0,0 +1,81 @@
+import torch
+
+# from brrr.utils import convert_tensor_to_fp8
+import transformer_engine as te  # noqa
+import transformer_engine_extensions as tex
+
+from nanotron.fp8.constants import DTypes
+from nanotron.fp8.meta import FP8Meta
+
+
+class FP8Tensor(torch.Tensor):
+    """FP8 Tensor."""
+
+    # TODO(xrsrke): add type hints for fp8_meta
+    def __new__(cls, tensor: torch.Tensor, dtype: DTypes):
+        # TODO(xrsrke): if the tensor is on cpu, then bypass the quantization
+        assert tensor.device != torch.device("cpu"), "FP8Tensor only supports CUDA device"
+        assert isinstance(dtype, DTypes)
+
+        inverse_scale = torch.tensor(1.0, device=tensor.device, dtype=torch.float32)
+        fp8_meta = FP8Meta(tensor.abs().max().clone(), dtype, inverse_scale)
+
+        if tensor.dtype not in [torch.int8, torch.uint8]:
+            fp8_tensor = convert_tensor_to_fp8(tensor, fp8_meta)
+        else:
+            fp8_tensor = tensor
+        fp8_meta._update_inverse_scale()
+        obj = torch.Tensor._make_subclass(cls, fp8_tensor)
+        obj._fp8_meta = fp8_meta
+        obj.fp8_meta = fp8_meta
+        obj.your_mom = True
+        return obj
+
+    # @property
+    # def fp8_meta(self) -> FP8Meta:
+    #     return self._fp8_meta
+
+    # @fp8_meta.setter
+    # def fp8_meta(self, value: FP8Meta):
+    #     self._fp8_meta = value
+
+    # def __repr__(self) -> str:
+    #     return f"FP8Tensor({self}, fp8_meta={self.fp8_meta})"
+
+
+def convert_torch_dtype_to_te_dtype(dtype: torch.dtype) -> tex.DType:
+    # setattr(torch, "fp8e5m2", torch.int8)
+    # setattr(torch, "fp8e4m3", torch.uint8)
+    from nanotron.fp8.constants import DTypes
+
+    # NOTE: transformer engine maintains it own dtype mapping
+    # so we need to manually map torch dtypes to te dtypes
+    TORCH_DTYPE_TE_DTYPE_NAME_MAPPING = {
```

Comment:
they don't have a mapping in the library themselves? 🥲

Reply:
They have this https://github.com/NVIDIA/TransformerEngine/blob/d68028c872186cd3df604e2ee2f60e09784d955c/transformer_engine/pytorch/cpp_extensions/fused_attn.py#L26. But it looks like a low-level detail. I'm afraid that if we depend on this and they refactor it out (since it's a low-level implementation detail), our code could break!!